flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/Splitter.scala

```diff
@@ -0,0 +1,210 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.preprocessing
+
+import org.apache.flink.api.common.typeinfo.{TypeInformation, BasicTypeInfo}
+import org.apache.flink.api.java.Utils
+import org.apache.flink.api.scala._
+import org.apache.flink.api.scala.DataSet
+import org.apache.flink.api.scala.utils._
+
+
+import org.apache.flink.ml.common.{FlinkMLTools, ParameterMap, WithParameters}
+import org.apache.flink.util.Collector
+import _root_.scala.reflect.ClassTag
+
+object Splitter {
+
+  case class TrainTestDataSet[T: TypeInformation : ClassTag](training: DataSet[T],
+                                                             testing: DataSet[T])
+
+  case class TrainTestHoldoutDataSet[T: TypeInformation : ClassTag](training: DataSet[T],
+                                                                    testing: DataSet[T],
+                                                                    holdout: DataSet[T])
+  // --------------------------------------------------------------------------------------------
+  //  randomSplit
+  // --------------------------------------------------------------------------------------------
+  /**
+   * Split a DataSet by the probability fraction of each element.
+   *
+   * @param input           DataSet to be split
+   * @param fraction        Probability that each element is chosen, should be [0,1] This fraction
+   *                        refers to the first element in the resulting array.
+   * @param precise         Sampling by default is random and can result in slightly lop-sided
+   *                        sample sets. When precise is true, equal sample set size are forced,
+   *                        however this is somewhat less efficient.
+   * @param seed            Random number generator seed.
+   * @return An array of two datasets
+   */
+
+  def randomSplit[T: TypeInformation : ClassTag](
+      input: DataSet[T],
+      fraction: Double,
+      precise: Boolean = false,
+      seed: Long = Utils.RNG.nextLong())
+    : Array[DataSet[T]] = {
+    import org.apache.flink.api.scala._
+
+    val indexedInput: DataSet[(Long, T)] = input.zipWithUniqueId
+
+    if ((fraction >= 1) || (fraction <= 0)) {
+      throw new IllegalArgumentException("sampling fraction outside of (0,1) bounds is nonsensical")
+    }
+
+    val leftSplit: DataSet[(Long, T)] = precise match {
+      case false => indexedInput.sample(false, fraction, seed)
+      case true => {
+        val count = indexedInput.count()
+        val numOfSamples = math.round(fraction * count).toInt
+        indexedInput.sampleWithSize(false, numOfSamples, seed)
+      }
+    }
+
+    val rightSplit: DataSet[T] = indexedInput.leftOuterJoin[(Long, T)](leftSplit)
```

Comment:
before doing the `leftOuterJoin` we could strip the `leftSplit` from it's payload and simply joining with the ids. That would reduce the network I/O.